eva/udfs/owl_object_detector.py

```diff
@@ -0,0 +1,230 @@
+import pandas as pd
+
+from eva.catalog.catalog_type import NdArrayType
+from eva.udfs.abstract.abstract_udf import AbstractUDF
+from transformers import OwlViTProcessor, OwlViTForObjectDetection
+from eva.utils.logging_manager import logger
+from eva.udfs.gpu_compatible import GPUCompatible
+from eva.udfs.decorators.decorators import forward, setup
+from eva.udfs.decorators.io_descriptors.data_types import PandasDataframe
+
+try:
+    import torch
+    from torch import Tensor
+
+except ImportError as e:
+    raise ImportError(
+        f"Failed to import with error {e}, \
+        please try `pip install torch`"
+    )
+
+
+class OwlObjectDetector(AbstractUDF, GPUCompatible):
+    """
+    Arguments:
+        threshold (float): Threshold for classifier confidence score
+
+    """
+
+    @property
+    def name(self) -> str:
+        return "yolo"
+
+    def to_device(self, device: str):
+        self.device = device
+        return self
+
+    @setup(cachable=True, udf_type="object_detection", batchable=True)
+    def setup(self, threshold=0.3):
+        logging.getLogger("hf_oneshot").setLevel(logging.CRITICAL)  # yolov5
+        self.threshold = threshold
+        self.processor = OwlViTProcessor.from_pretrained("google/owlvit-base-patch32")
+        self.model = OwlViTForObjectDetection.from_pretrained("google/owlvit-base-patch32")
+        # self.model = torch.hub.load("ultralytics/yolov5", "yolov5s", verbose=False)
+        self.device = "cpu"
+
+    @forward(
+        input_signatures=[
+            PandasDataframe(
+                columns=["data"],
+                column_types=[NdArrayType.FLOAT32],
+                column_shapes=[(None, None, 3)],
+            )
+        ],
+        output_signatures=[
+            PandasDataframe(
+                columns=["labels", "bboxes", "scores"],
+                column_types=[
+                    NdArrayType.STR,
+                    NdArrayType.FLOAT32,
+                    NdArrayType.FLOAT32,
+                ],
+                column_shapes=[(None,), (None,), (None,)],
+            )
+        ],
+    )
+    def forward(self, frames: pd.DataFrame) -> pd.DataFrame:
+        """
+        Performs predictions on input frames
+        Arguments:
+            frames (np.ndarray): Frames on which predictions need
+            to be performed
+        Returns:
+            tuple containing predicted_classes (List[List[str]]),
+            predicted_boxes (List[List[BoundingBox]]),
+            predicted_scores (List[List[float]])
+        """
+        # Stacking all frames, and changing to numpy
+        # because of yolov5 error with Tensors
+        print("Does this work?")
+        outcome = []
+        # Convert to HWC
+        # https://github.com/ultralytics/yolov5/blob/3e55763d45f9c5f8217e4dad5ba1e6c1f42e3bf8/models/common.py#L658
+        frames = np.ravel(frames.to_numpy())
+        frames = torch.permute(frames, (0, 2, 3, 1))
+        logger.warn("Does this work?1")
```

Comment:
These logger.warn lines merely check if each line works during initial debugging. Perhaps they should be deleted?